### PR TITLE
KEYCLOAK-6698 - Add substitution of system properties and environment variables in theme.properties file

### PR DESCRIPTION
**Pull request by Opa-**
Hello everyone,

Did the work for this issue https://issues.jboss.org/browse/KEYCLOAK-6698
Although I do not know the procedure to re-open the ticket.

Regards

### CODE REVIEW

**Comment by Opa-** [comment]
Test is failing due to missing dependency on commons-lang package from Apache.

I though the dependency was already there but I guess my IDE tricked me as it is only used by the tests.

As per the following document https://github.com/keycloak/keycloak/blob/master/docs/building.md the addition of dependency should be discussed in the mailing list before. Do I need to start a discussion on the mailing list or is it ok to discuss it here ?

**Comment by stianst** [comment]
Take a look at org.keycloak.common.util.StringPropertyReplacer#replaceProperties(java.lang.String, java.util.Properties) and also org.keycloak.common.util.SystemEnvProperties.

File: services/src/main/java/org/keycloak/theme/ExtendingThemeManager.java [hunk 1]
```diff
@@ -172,6 +176,8 @@ private Theme findTheme(String name, Theme.Type type) {
     }
 
     public static class ExtendingTheme implements Theme {
+        private final static Pattern ENVIRONMENT_VARIABLE_SUBSTITUTE_PATTERN = Pattern.compile("\\$\\{env.([A-Za-z0-9_.-]+)(?::([^\\}]*))?\\}");
```
Comment (stianst) [inline services/src/main/java/org/keycloak/theme/ExtendingThemeManager.java]: This code doesn't belong here. Take a look at:
* org.keycloak.common.util.SystemEnvProperties
* org.keycloak.common.util.StringPropertyReplacer

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/theme/ExtendingThemeTest.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        testingClient.server().run(session -> {
+            try {
+                ThemeProvider extending = session.getProvider(ThemeProvider.class, "extending");
+                Theme theme = extending.getTheme(THEME_NAME, Theme.Type.LOGIN);
+                Assert.assertEquals("", theme.getProperties().getProperty("env.missing"));
+                Assert.assertEquals("defaultValue", theme.getProperties().getProperty("env.missingWithDefault"));
+                if (System.getenv().containsKey("HOMEPATH")) {
+                    // Windows
+                    Assert.assertEquals(System.getenv().get("HOMEPATH"), theme.getProperties().getProperty("env.windowsHome"));
+                } else if (System.getenv().containsKey("HOME")) {
```
Comment (stianst) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/theme/ExtendingThemeTest.java]: Need to confirm if this is set on Mac

**Review by stianst — CHANGES_REQUESTED** [review]

File: services/src/main/java/org/keycloak/theme/ExtendingThemeManager.java (hunk 1 above)
Comment (Opa-) [reply services/src/main/java/org/keycloak/theme/ExtendingThemeManager.java]: Ok thanks for the advice ! 
With these two classes I got rid of the dirty things I added to ExtendingTheme class 👍

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/theme/ExtendingThemeTest.java (hunk 2 above)
Comment (Opa-) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/theme/ExtendingThemeTest.java]: `$HOME` seems to be ok for most of the other OS https://en.wikipedia.org/wiki/Home_directory
There's no official doc for mac, but some of my coworkers uses macs and it is set.

**Review by stianst — APPROVED** [review]

**Comment by stianst** [comment]
Nice job - one last thing would you mind sending a PR to https://github.com/keycloak/keycloak-documentation/blob/master/server_development/topics/themes.adoc to mention that theme.properties supports replacing system propeties and env variables? If we don't add to docs no-one except me and you will now it exists

**Comment by Opa-** [comment]
Thanks for the review

PR submitted for the doc https://github.com/keycloak/keycloak-documentation/pull/746 😉

**Comment by Opa-** [comment]
I don't know if it's important but the related ticket https://issues.jboss.org/browse/KEYCLOAK-6698 is currently in CLOSED status and unassigned (because out of date)

**Comment by stianst** [comment]
I've opened the JIRA again, it was just closed due to inactivity.

**Review by stianst — APPROVED** [review]

**Pull request merged**